### DISCUSSION TITLE
Github should support closing issues by merging PRs to non-main branches

### DISCUSSION THREAD

**Discussion by xmo-odoo** [community/community#112224] [Pull Requests]
For a long time now github has supported closing issues [via commit messages](https://github.blog/2013-01-22-closing-issues-via-commit-messages/) and [pull request descriptions](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

However this suffers from a very frustrating limitation: it only works when the commit lands in or the PR targets the repository's default branch.

This means it's almost completely useless for projects with long running / maintenance branch, because bug fix PRs will commonly target non-main branches, and the commits might not even land in the main branch ever, depending on the project's methodology (merge v back-port v forward-port).

# Example

[The CPython project](https://github.com/python/cpython/) has yearly major releases and supports each major for about 5 (five) years, ignoring 2.7 which got 10 years of support. Their maintenance policy is to fix the latest branch in which an issue occurs, then have [a bot backport the fix to older branches](https://github.com/python/miss-islington). So e.g. if a fix happens in 3.10 but not in 3.11 (because the subsystem was rewritten or removed), a PR will be created against 3.10 and the commit will land in 3.10, 3.9, and 3.8, but not the default branch (main), nor branches for later major releases (3.12 and 3.11).

# Request

It should be possible to mark non-main branches as keyword targets / whatever. I believe the *rulesets* could be leveraged for that, as usually a branch protected by a ruleset would be a primary or relevant branch (rather than a development scratch branch), and thus might benefit from automatically closing issues.

**Comment by moozunch** [comment]
Hi, I think when you open PR and close it right after it will remain in that branch without moving to main branch, or I got the question wrong?

**Reply by xmo-odoo** [reply]
I'm not too sure how you interpreted the request, or how to make it clearer. It has nothing to do with closing PRs, it has to do with closing *issues* via PRs.

Let me explain a demo: create a repository with a default branch. Create 3 additional branches B, C, and D (they all have the same content, doesn't matter)

Create two issues #1 and #2.

Go to branch C, create a new change, then create a PR #3 targeting the default branch, in the description of the PR enter "fixes #1".

Go to branch D, create a new change, then create a PR #4 targeting B, in the description of the PR enter "fixes #2".

In the side-bar of the first PR (#3), under "Development" you should see something like

> Successfully merging this pull request may close these issues
> (whatever title you gave to issue 1)

In the side-bar of the second PR (#2) you will not see anything like it, despite the similar message. Still you can click on the gear and select issue 2 to get a similar display to the first PR.

Merge the two PRs.

Notice that the issue you linked to the first PR (targeting the main branch) is now closed.

Notice that the issue you linked to the second PR (targeting an other branch) is *not* closed.

I want the ability to close issues from PRs targeting non-default branches.

**Comment by Alex-ley-scrub** [comment]
We need this so badly - how is it possible in the UI but not via GitHub Rest/GraphQL API?!

see meta issue: https://github.com/orgs/community/discussions/186495
